### PR TITLE
Fix #161: Exploration player contentcard supports rich-text part  [Blocked: #205]

### LINKED ISSUE #161: UI Structure: ContentCardView
This is tracking implementing the initial structure for the content card view that's being fully implemented in #26. This doesn't need to support rich-text support yet, but at least having a way to process Oppia's custom HTML tags seems worthwhile since it may be complex and require infrastructural work.

This also doesn't need to have any special data associated with it (hence why it has no domain controller blockers).

### PR DESCRIPTION
**Pull request by veena14cs**
##  Explanation
This PR depends on #205 which is under review and base off merge-fix which contains code of #205. 
This PR is mainly to focus on exploration content card it contains implementation of html parser  which parses HTML content. Even if the implementation of #205 changes it does not effect this PR.

Recyclerview Matcher file is added in this PR for testcase, It is already present in develop. Once  #205  is approved and update to develop this file will be removed.

### CODE REVIEW

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateAdapter.kt]: Incorrect spacing. Update this simple by pressing: `Option+command+L`

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:70]: Remove this extra line

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: I think that this `explorationId` will not change, so better make it `val` and also use `lateinit`with this.

**Review by rt4914 — APPROVED** [review]
Generally LGTM.

Do we need to introduce test cases in this PR?

**Comment by rt4914** [comment]
@veena14cs I suggest writing description related to the `merge-fix` branch and what does it contain and why are you using this, both of us has communicated regarding this but Ben and Nikita are not aware about this.

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateAdapter.kt]: corrected.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:70]: removed extra lines.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: I am initializing it in onCreate(). so defined it var.

**Comment by veena14cs** [comment]
> Generally LGTM.
> 
> Do we need to introduce test cases in this PR?

Test case related to what actually.

**Review by BenHenning — APPROVED** [review]
Thanks @veena14cs. Why is this marked as do not merge? Is it just that it's blocked on #205? If so, please update the title to reflect it's blocked. "Do not merge" to me feels like it should never be merged, but in this case it's unclear.

I agree with @rt4914 that we should add a test case for this. Perhaps one test in StateFragment to verify it properly adds spannables to state with Oppia HTML content. That's expected behavior for StateFragment, and that test should fail if you removed the code being added here. That's then a worthwhile test since it's verifying the integration works.

Once that test is added, this change LGTM.

**Comment by rt4914** [comment]
> Thanks @veena14cs. Why is this marked as do not merge? Is it just that it's blocked on #205? If so, please update the title to reflect it's blocked. "Do not merge" to me feels like it should never be merged, but in this case it's unclear.
> 
> I agree with @rt4914 that we should add a test case for this. Perhaps one test in StateFragment to verify it properly adds spannables to state with Oppia HTML content. That's expected behavior for StateFragment, and that test should fail if you removed the code being added here. That's then a worthwhile test since it's verifying the integration works.
> 
> Once that test is added, this change LGTM.

@BenHenning Actually, do not merge was added by me so that by mistake someone should not merge this because the `merge-fix` is a combination of different branches which have been not approved yet. I will check with @veena14cs on which one it is blocked an fix it accordingly.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: Suggestions:

`testContentCard_loadExplorationTest5_handleCustomOppiaTags_parsedHtmlDisplaysCorrectly`

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: Remove `ViewActions` and import it directly.

File: app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: Add a comment and also mention in descroiption that why this file is in this PR even though the file already exists in `develop`.

**Review by rt4914 — COMMENTED** [review]
PTAL

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: corrected.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: done.

File: app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: Added comment. This file is not directly from develop its from merge-fix. Once merge-fix is updated from develop this file will be removed.

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: @veena14cs Remove this comment from here, apologies I was not clear in earlier comment.

I meant at comment on Github,I mentioned that in "Files Changed" and also, mention that in PR description and not in this file. Because we do not want this comment in develop code.

**Review by rt4914 — APPROVED** [review]
LGTM.
Nit Change PTAL

File: app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: Okay, removed it from code.

**Comment by BenHenning** [comment]
@veena14cs do I need to take another look at this? If so, what do you specifically want me to look at? The PR looked fine to me yesterday, so if my previous comments were resolved then feel free to submit. Otherwise I'm happy to look at any parts you'd like me to again.

**Comment by veena14cs** [comment]
> @veena14cs do I need to take another look at this? If so, what do you specifically want me to look at? The PR looked fine to me yesterday, so if my previous comments were resolved then feel free to submit. Otherwise I'm happy to look at any parts you'd like me to again.

Yes, please take a glance at testcase.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Got it. with `lateinit` `val` is not possible.

**Comment by BenHenning** [comment]
@veena14cs please resolve the existing conflicts in this PR then send it back to me.

**Comment by veena14cs** [comment]
> @veena14cs please resolve the existing conflicts in this PR then send it back to me.

@BenHenning Resolved conflicts.

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/StateAdapter.kt:35]: Do we need to pass this in if we know that the entity ID is an exploration ID? I'm guessing we can infer at this point that the entity type will be exploration.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: It actually might be nice to inject this, instead. Then we could have a qualifier bind the value to the Dagger graph to keep its definition in one location, e.g.:

Qualifier:
```kotlin
@Qualifier annotation class ExplorationHtmlParserEntityType
```

Module:
```kotlin
@Module
class HtmlParserEntityTypeModule {
  @Provides
  @ExplorationHtmlParserEntityType
  fun provideExplorationHtmlParserEntityType(): String {
    return "exploration"
  }
}
```

Everywhere where we currently define this, instead inject:
```kotlin
class ExampleClass @Inject(@ExplorationHtmlParserEntityType private val entityType: String) {
  ...
}
```

In general I think we should aim to keep sentinel values like this in the Dagger graph to make it easier to share them everywhere and minimize the chance of accidentally getting it wrong in one spot (or needing to update all spots if the value needs to change in the future).

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: What does ``loadExplorationTest5`` correspond to? Is it that a specific exploration is being loaded?

Maybe just: ``testContentCard_forDemoExploration_withCustomOppiaTags_displaysParsedHtml()``

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt:20]: Please make sure in a subsequent PR that we also test the image loading case here.

File: app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: Why do we need this class? Isn't it already in ``develop``? (https://github.com/oppia/oppia-android/blob/develop/app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt).

Please remove (or merge ``develop`` into your base branch to bring this up-to-date).

**Review by BenHenning — APPROVED** [review]
Thanks @veena14cs. This LGTM!

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateAdapter.kt:35]: To use it in HtmlParser, that is reusable class. I am specifying `entityType` to "exploration" because HtmlParser takes it as parameter to load images ,if its not from exploration and any other entity we can pass that string.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: Changed test name to `testContentCard_forDemoExploration_withCustomOppiaTags_displaysParsedHtml()`

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Done.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt:20]: Okay will add it once it is ready.

File: app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt]: Recyclerview Matcher file is added in this PR for testcase, It is already present in develop. Once #205 is approved and update to develop this file will be removed.

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateAdapter.kt:35]: Injected entityType.

File: app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/application/ApplicationComponent.kt]: Please wrap these if they exceed the line limit.

File: app/src/main/java/org/oppia/app/player/state/HtmlParserEntityTypeModule.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/HtmlParserEntityTypeModule.kt]: Suggest putting this next to the HtmlParser in the same package.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt:20]: Thanks. I suggest adding a TODO in the interim as a reminder in code.

**Review by BenHenning — APPROVED** [review]
Thanks @veena14cs! This LGTM!

File: app/src/main/java/org/oppia/app/player/state/HtmlParserEntityTypeModule.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/HtmlParserEntityTypeModule.kt]: Done.

File: app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/application/ApplicationComponent.kt]: Done.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt:20]: Added.

**Comment by rt4914** [comment]
@veena14cs As per @BenHenning last comments I see that you were suppose to make `entityType` injectable and you have done that. 

So this PR still LGTM.

De-assigning myself.

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (nikitamarysolomanpvt) [inline app/src/main/java/org/oppia/app/player/state/StateAdapter.kt:66]: Please remove the use less cast in line 67

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (nikitamarysolomanpvt) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Remove Empty line

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (nikitamarysolomanpvt) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: Please remove all unused imports

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM suggested some nit changes PTL

File: app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateAdapter.kt:66]: done.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: removed.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt]: removed.

**Review by BenHenning — CHANGES_REQUESTED** [review]
This branch is not merge compatible with develop. Please make this based off of develop before merging. I believe there are hidden changes in this branch that haven't yet been reviewed after cerateing #265.

**Comment by veena14cs** [comment]
Duplicate of this PR is create here #275 .

**Pull request closed without merging**
